[mlir] Guard expensive string-based asserts behind EXPENSIVE_CHECKS

Mogball opened the pull request:
These string based asserts are important but they are extremely expensive. In compilation workloads with lots of interface checking, this can slow down as much as 3x.

joker-eph commented on mlir/include/mlir/IR/Attributes.h:
```diff
@@ -50,20 +50,17 @@ class Attribute {
   template <typename U>
-  [[deprecated("Use mlir::cast<U>() instead")]]
-  U cast() const;
+  [[deprecated("Use mlir::cast<U>() instead")]] U cast() const;
```
Unrelated formatting change?

Mogball replied:
clang-format strikes again!

joker-eph commented on mlir/include/mlir/IR/Attributes.h:
```diff
@@ -303,7 +300,7 @@ class AttributeInterface
   /// Returns the impl interface instance for the given type.
   static typename InterfaceBase::Concept *getInterfaceFor(Attribute attr) {
-#ifndef NDEBUG
+#ifdef EXPENSIVE_CHECKS
```
This is critical to check in asserts builds IMO. The promise is quite useless otherwise...

Mogball replied:
I'm open to your opinion on this! On one hand, I agree. I opened this PR simply for discussion (and to evaluate the impact on mojo). Practically no one builds with EXPENSIVE_CHECKS by default. However, compared to asserts in, say, `isa` or `cast`, this one is realllyyyy expensive. What if we had a separate flag that was also on by default but toggleable? The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. It would be nice to specifically toggle these ones and other "heavy" asserts in MLIR off.

Mogball replied:
Our compiler(s) are under development, so turning off asserts in downstream projects might not be as desirable. Having a slow compiler, however, is also affecting the developer experience.

joker-eph replied:
> What if we had a separate flag that was also on by default but toggleable?

That seems fine to me.

> The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. 

Which is WAI I believe: they build in assert mode means they want to catch issues in their project, so this will catch issues with their uses of isa/...

stellaraccident replied:
It's a balance. Subjectively, I think that asserts in milr are already over the line of being too expensive, but it is a fraught topic to bring up. I can't say whether this is one of the big culprits, but expensive checks in isa costing 3x of overall compile time is both troubling and would explain a lot based on my experience.

joker-eph reviewed (changes requested):
I have strong concerns here about making the system unsafe and much harder to debug/develop.

Commit:
[mlir][ods] Optimize FoldAdaptor constructor

FoldAdaptor is generated as a subclass of the operation's generic
adaptor, which requires an OperationName instance. It called into the
generic base constructor that constructed the OperationName from a
string, requiring a StringMap lookup inside the MLIRContext.

This makes constructing FoldAdaptors really slow, which is a shame
because the `Operation *` is right there. This PR changes GenericAdaptor
constructor from an operation to grab the OperationName directly from
the `Operation *`. In addition, it generates the constructor inline if
the operation doesn't have properties, since otherwise it requires the
definition of the op.

Mogball commented:
I might come back to this one with a different set of flags.

joker-eph commented:
Thanks Jeff! Having more flags to `mlir-config.h` seems like a good direction to me!
